### PR TITLE
feat(analytics): Allow sending batch events

### PR DESCRIPTION
**Pull request by iartemiev**
_Issue #, if available:_
Fixes: #1989

_Description of changes:_
**Enables batch event sending.**

I took out the other features I'm working on as part of the Analytics Refactor because I need to switch over to another task for a little bit and wanted to get this out sooner rather than later.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/analytics/__tests__/Providers/AWSPinpointProvider-unit-test.ts [hunk 1]
```diff
@@ -501,7 +501,8 @@ describe('AnalyticsProvider test', () => {
 				await analytics.record(params, { resolve, reject });
 
 				const expectedUrl =
-					'https://pinpoint.region.amazonaws.com/v1/apps/appId/events?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=accessKeyId%2FisoStrin%2Fregion%2Fmobiletargeting%2Faws4_request&X-Amz-Date=isoString&X-Amz-Security-Token=sessionToken&X-Amz-SignedHeaders=host&X-Amz-Signature=c43336d302010144ca492cc6f42d5545e0ef01c00bc71d99dc7597351f027810';
+					'https://pinpoint.region.amazonaws.com/v1/apps/appId/events/legacy?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=accessKeyId%2FisoStrin%2Fregion%2Fmobiletargeting%2Faws4_request&X-Amz-Date=isoString&X-Amz-Security-Token=sessionToken&X-Amz-SignedHeaders=host&X-Amz-Signature=9dfa2a29782d344c56a9ab99fe58db6d1748e097ae418c398b26ab372a23f22f';
```
Comment (sammartinez) [inline packages/analytics/__tests__/Providers/AWSPinpointProvider-unit-test.ts]: Can we confirm that `Amz-Credential=accessKeyId%2FisoStrin%2Fregion` is not a typo on the `Strin`? Im just seeing `isoString` in other places

File: packages/analytics/src/Analytics.ts [hunk 2]
```diff
@@ -254,7 +254,7 @@ export default class AnalyticsClass {
 	}
 
 	public async updateEndpoint(attrs, provider?) {
-		const event = Object.assign({ name: '_update_endpoint' }, attrs);
+		const event = { ...attrs, name: '_update_endpoint' };
```
Comment (sammartinez) [inline packages/analytics/src/Analytics.ts]: Will this work? We may not have this compile down just want to double check

File: packages/analytics/src/Providers/AWSPinpointProvider.ts [hunk 3]
```diff
@@ -415,36 +619,35 @@ export default class AWSPinpointProvider implements AnalyticsProvider {
...
+			logger.debug('updateEndpoint success', data);
+			this._endpointGenerating = false;
+			return handlers.resolve(data);
+		} catch (err) {
+			logger.debug('updateEndpoint failed', err);
+			if (err.message.startsWith('Exceeded maximum endpoint per user count')) {
+				try {
+					await this._removeUnusedEndpoints(appId, request.User.UserId);
+
+					logger.debug('Remove the unused endpoints successfully');
```
Comment (sammartinez) [inline packages/analytics/src/Providers/AWSPinpointProvider.ts]: Remove or Removed?

File: packages/analytics/src/Providers/AWSPinpointProvider.ts [hunk 4]
```diff
@@ -644,12 +847,26 @@ export default class AWSPinpointProvider implements AnalyticsProvider {
...
+	private _bufferToMap(buffer: EventBuffer) {
+		return buffer.reduce((acc, curVal) => {
+			const [[key, value]] = Object.entries(curVal);
+			acc[key] = value;
+			return acc;
+		}, {});
+	}
+
+	private _eventError(err: any) {
+		logger.error('record event failed. ', err);
```
Comment (sammartinez) [inline packages/analytics/src/Providers/AWSPinpointProvider.ts]: Is this space needed after `record event failed. `?

File: packages/core/src/Credentials.ts [hunk 5]
```diff
@@ -158,7 +158,7 @@ export class Credentials {
 		}
 		logger.debug('is this credentials expired?', credentials);
 		const ts = new Date().getTime();
-		const delta = 10 * 60 * 1000; // 10 minutes
+		const delta = 10 * 60 * 1000 - 5000; // 10 minutes minus 5 seconds
```
Comment (sammartinez) [inline packages/core/src/Credentials.ts]: why the change? Just curious is all. Related to the batch of events?

File: packages/analytics/src/Analytics.ts (hunk 2 above)
Comment (iartemiev) [reply packages/analytics/src/Analytics.ts]: It'll work. We use object prop destructuring elsewhere as well. 
For example: https://github.com/aws-amplify/amplify-js/blob/master/packages/api/src/API.ts#L68

File: packages/core/src/Credentials.ts (hunk 5 above)
Comment (iartemiev) [reply packages/core/src/Credentials.ts]: It's not, it's related to the Creds / Retry stuff I'm also working on

Comment (Amplifiyer) [reply packages/core/src/Credentials.ts]: @elorzafe is also working on something similar (if not same). Let's sync up

Comment (iartemiev) [reply packages/core/src/Credentials.ts]: I'll take it out for now though

File: packages/analytics/src/Providers/AWSPinpointProvider.ts [hunk 6]
```diff
@@ -79,34 +90,73 @@ export default class AWSPinpointProvider implements AnalyticsProvider {
...
+			const endpointBatchSize =
+				that._endpointBuffer.length < flushSize
+					? that._endpointBuffer.length
+					: flushSize;
+
+			if (!batchSize && !endpointBatchSize) {
+				return;
+			}
+
+			// TODO: figure out if this is still relevant
```
Comment (jordanranz) [inline packages/analytics/src/Providers/AWSPinpointProvider.ts]: Is this TODO still needed?

Comment (iartemiev) [reply packages/analytics/src/Providers/AWSPinpointProvider.ts]: Probably not haha. I'll remove it

File: packages/analytics/src/Providers/AWSPinpointProvider.ts [hunk 7]
```diff
@@ -79,34 +90,72 @@ export default class AWSPinpointProvider implements AnalyticsProvider {
...
+			}
+
+			if (that._endpointGenerating && endpointBatchSize) {
 				// If this is the first request sent by Analytics module, we should stop sending remaining requests
 				// to prevent race condition of updating one endpoint when it's being created in the backend
-				if (this._endpointGenerating) break;
+				const firstEvent = that._endpointBuffer.shift();
+				const [{ params, handlers }] = Object.values(firstEvent);
+				that._send(params, handlers);
+				return;
```
Comment (elorzafe) [inline packages/analytics/src/Providers/AWSPinpointProvider.ts]: that will wait for `flushInterval` to do the next one right?, what happen if customer set that to low value by default. How long should it wait, in case the app sends an immediate event it should wait for that, or maybe there should an exponential retry for an error because of waiting for first update endpoint event?

File: packages/analytics/src/Providers/AWSPinpointProvider.ts [hunk 8]
```diff
@@ -389,15 +565,39 @@ export default class AWSPinpointProvider implements AnalyticsProvider {
...
+
+		const batchEndpointUpdates = this._batchEndpoints(endpointsMap);
+		logger.debug('batchEndpointUpdates', batchEndpointUpdates);
+
+		try {
+			const data = await this.pinpointClient
+				.updateEndpointsBatch(batchEndpointUpdates)
+				.promise();
+
+			// TODO: handle response
```
Comment (elorzafe) [inline packages/analytics/src/Providers/AWSPinpointProvider.ts]: I think you are missing something here ;)

File: packages/analytics/src/Providers/AWSPinpointProvider.ts [hunk 9]
```diff
@@ -79,34 +90,72 @@ export default class AWSPinpointProvider implements AnalyticsProvider {
...
 		const that = this;
-		this._timer = setInterval(() => {
-			const size =
-				this._buffer.length < flushSize ? this._buffer.length : flushSize;
-			for (let i = 0; i < size; i += 1) {
-				const { params, handlers } = this._buffer.shift();
-				that._send(params, handlers);
+
+		this._timer = setInterval(async () => {
+			await that._getCredentials(); // TODO: set class creds here
```
[the remark this answers is not in the document]
Comment (elorzafe) [reply packages/analytics/src/Providers/AWSPinpointProvider.ts]: There might be a refactor on `_initClients` method, because the credentials are not store on this provider

File: packages/analytics/src/Providers/AWSPinpointProvider.ts [hunk 10]
```diff
@@ -369,6 +544,7 @@ export default class AWSPinpointProvider implements AnalyticsProvider {
 	}
 
 	private _retry(params, handlers) {
+		logger.debug('_retry', params, handlers);
```
Comment (elorzafe) [inline packages/analytics/src/Providers/AWSPinpointProvider.ts]: I think this retry logic has a bug since a long time, is putting the retryable events on the end of the buffer, or maybe it shouldn't be an issue. What do you think?

File: packages/analytics/src/Providers/AWSPinpointProvider.ts [hunk 11]
```diff
@@ -79,34 +90,72 @@ export default class AWSPinpointProvider implements AnalyticsProvider {
...
 		if (this._buffer.length < bufferSize) {
-			this._buffer.push({ params, handlers });
+			const {
+				event: { eventId, name },
+			} = params;
+
+			const bufferElement = { [eventId]: { params, handlers } };
+
+			if (name === UPDATE_ENDPOINT) {
+				this._endpointBuffer.push(bufferElement);
```
Comment (elorzafe) [inline packages/analytics/src/Providers/AWSPinpointProvider.ts]: this could be an issue with `updateEndpoint`, for retry it will put it on the end of the buffer an can override other updateEndpoints (basically processing in sequence it will be broken)

**Comment by iartemiev** [comment]
Closed this PR and re-opened as #4377 due to unintentionally getting my local branch too far out of sync with this one through rebasing latest master, etc.
